src/vs/workbench/api/node/mainThreadLanguageFeatures.ts

```diff
@@ -134,6 +134,28 @@ export class MainThreadLanguageFeatures extends MainThreadLanguageFeaturesShape
 		return undefined;
 	}
 
+	// --- external references
+
+	$registerExternalReferenceSupport(handle: number, selector: vscode.DocumentSelector, workspace?: IWorkspace): TPromise<any> {
+		this._registrations[handle] = modes.ExternalReferenceProviderRegistry.register(selector, <modes.ExternalReferenceProvider>{
+			provideExternalReferences: (query: modes.ISymbolDescriptor, hints: { [hint: string]: any }, token: CancellationToken, progress: (references: modes.IReferenceInformation[]) => void): modes.IReferenceInformation[] | Thenable<modes.IReferenceInformation[]> => {
+				const progressHandle = this._callbackRegistrations.registerChild(handle, progress);
+				let refs = this._proxy.$provideExternalReferences(handle, progressHandle, query, hints);
+				refs = always(refs, () => this._callbackRegistrations.unregisterChild(handle, progressHandle));
+				return wireCancellationToken(token, refs);
+			}
+		}, workspace);
+		return undefined;
```

Comment:
I am pretty sure the other code was written before TypeScript supported `undefined` types and strict null checks and they just didn't refactor everything. Imo new code should be typed correctly.